core/src/validator.rs

```diff
@@ -2181,6 +2183,63 @@ fn maybe_warp_slot(
     Ok(())
 }
 
+/// Returns the starting slot at which the blockstore should be scanned for
+/// shreds with an incorrect shred version, or None if the check is unnecessary
+fn should_check_blockstore_for_incorrect_shred_version(
+    config: &ValidatorConfig,
+    blockstore: &Blockstore,
+    root_slot: Slot,
+    hard_forks: &HardForks,
+) -> Result<Option<Slot>, BlockstoreError> {
+    // Perform the check if we are booting as part of a cluster restart at slot root_slot
+    let maybe_cluster_restart_slot = maybe_cluster_restart_with_hard_fork(config, root_slot);
+    if maybe_cluster_restart_slot.is_some() {
+        return Ok(Some(root_slot + 1));
+    }
+
+    // If there are no hard forks, the shred version cannot have changed
+    let Some(latest_hard_fork) = hard_forks.iter().last().map(|(slot, _)| *slot) else {
+        return Ok(None);
+    };
+
+    // If the blockstore is empty, there are certainly no shreds with an incorrect version
+    let Some(blockstore_max_slot) = blockstore.highest_slot()? else {
+        return Ok(None);
+    };
+    let blockstore_min_slot = blockstore.lowest_slot();
+
+    if latest_hard_fork < blockstore_min_slot {
+        // latest_hard_fork < blockstore_min_slot <= blockstore_max_slot
+        //
+        // All slots in the blockstore are newer than the latest hard fork, and only shreds with
+        // the correct shred version should have been inserted since the latest hard fork
+        //
+        // This is the normal case where the last cluster restart & hard fork was a while ago; we
+        // can skip the check for this case
+        Ok(None)
+    } else if latest_hard_fork < blockstore_max_slot {
+        // blockstore_min_slot < latest_hard_fork < blockstore_max_slot
+        //
+        // This could be a case where there was a cluster restart, but this node was not part of
+        // the supermajority that actually restarted the cluster. Rather, this node likely
+        // downloaded a new snapshot while retaining the blockstore, including slots beyond the
+        // chosen restart slot. We need to perform the blockstore check for this case
+        //
+        // Note that the downloaded snapshot slot (root_slot) could be greater than the latest hard
+        // fork slot. Even though this node will only replay slots after root_slot, start the check
+        // at latest_hard_fork + 1 to check (and possibly purge) any invalid state.
+        Ok(Some(latest_hard_fork + 1))
+    } else {
+        // blockstore_min_slot <= blockstore_max_slot <= latest_hard_fork
+        //
+        // This could be similar to the previous case of there being a cluster restart, except
+        // that the blockstore data is all older than latest hard fork (perhaps this node crashed
+        // before observing the slot chosen for cluster restart). Or, maybe there was a hard fork
+        // issued very far into the future. Regardless, perform the blockstore check for this case.
```

Comment:
Is there a reason to perform the scan here? If I understand correctly, we always have a slot meta if a shred was inserted for that slot. Since this is the highest slot meta, that means there should be no shreds past the latest hard fork.

Reply:
> Is there a reason to perform the scan here? If I understand correctly, we always have a slot meta if a shred was inserted for that slot. Since this is the highest slot meta, that means there should be no shreds past the latest hard fork.

Good callout, in an earlier commit, the return value here could have been `root_slot + 1`. Given
```
bstore_min_slot <= bstore_max_slot    AND
root_slot < latest_hard_fork          AND
bstore_max_slot <= latest_hard_fork
```
it could have been the case that:
```
bstore_min_slot < root_slot < bstore_max_slot < latest_hard_fork
```
But, this has since been updated. If we know that all data is older than latest_hard_fork and that scan would start beyond latest_hard_fork, the operation would essentially be a no-op but it is more clear to skip the scan altogether. Thanks for catching, will update.

Reply:
https://github.com/anza-xyz/agave/pull/2851/commits/3c3e7c36575cdad565c4a0324481d364b8bc9da6